### ISSUE TITLE
CI is using an incompatible version of the Conda runtime

### ISSUE THREAD

**Issue by victorlin** [nextstrain/monkeypox#177]
Currently (as observed in #176), the Conda runtime job instance of `pathogen-ci` is [failing](https://github.com/nextstrain/monkeypox/actions/runs/6029807052/job/16360195382?pr=176#step:8:11) with the following error:

```
Current augur version: 22.1.0. Minimum required: 22.2.0
```

Augur version 22.1.0 is coming from [this version](https://github.com/nextstrain/monkeypox/actions/runs/6029807052/job/16360195382#step:6:1564) of the Conda runtime: `nextstrain-base 20230717T174555Z`.

This used to work without any noticeable changes. Example: when the Augur minimum version was bumped to 22.2.0, Augur version 22.2.0 was available in [this CI run](https://github.com/nextstrain/monkeypox/actions/runs/5732108745/job/15534404912#step:6:1479). Notably, the version of the Conda runtime is `nextstrain-base 20230731T212806Z`.

This also seems to be working fine in the ncov repo, where [the latest run](https://github.com/nextstrain/ncov/actions/runs/6028247945/job/16355149569#step:6:1480) resolved to `nextstrain-base 20230830T164409Z`.

My outstanding question is: why is an older version of the Conda runtime being resolved now, and seemingly only in this repo?

**Comment by victorlin** [comment]
Hmm, that's helpful info. It doesn't explain the behavior in the [ncov run](https://github.com/nextstrain/ncov/actions/runs/6028247945/job/16355149569#step:6:1480) though? That one resolved to `nextstrain-base 20230830T164409Z` during `nextstrain setup conda`.

**Comment by tsibley** [comment]
Yeah. Weird.

**Comment by joverlee521** [comment]
Seeing the same issue in the [seasonal-flu CI](https://github.com/nextstrain/seasonal-flu/actions/runs/6052311874/job/16425512518#step:6:1564) now, so no longer just limited to this repo.

**Comment by tsibley** [comment]
I can reproduce this locally even with the standalone install of Nextstrain CLI by setting up a new Conda runtime from scratch, which makes sense given we think this is a Micromamba issue.

**Comment by joverlee521** [comment]
[Quick fix](https://github.com/nextstrain/.github/pull/48) for the CI while we figure out the underlying issue.

**Comment by tsibley** [comment]
Thanks for the hot fix for CI!

I started digging into what's going on inside Micromamba by doing roughly this:

```console
$ cd $(mkdir -dt)
$ export NEXTSTRAIN_HOME=$PWD
$ nextstrain debugger
(Pdb) interact
>>> from nextstrain.cli.runner.conda import micromamba, setup_micromamba
>>> setup_micromamba()
>>> micromamba("create", "-vv", "--dry-run", "nextstrain-base")
…
```

I confirmed that the package index it's using, https://conda.anaconda.org/nextstrain/linux-64/repodata.json, contains the latest package version. It does. Then I [diffed the two index entries](https://gist.github.com/tsibley/db4fe60ddced250e9420d980212cddbb) to see if anything stood out, but nothing does.

Next to dig into the actual solver logs.

**Comment by tsibley** [comment]
The solver starts by considering the latest version, `20230830T164409Z`. It finds some conflict when solving deps between the `suitesparse 5.10.1` and `metis 5.1.1` packages, even though it should be fine to just install the exact versions listed in the nextstrain-base spec. That conflict is resolved by the solver by ruling out `20230830T164409Z` and repeating the process with the next highest version all the way down the line until it gets to `20230717T174555Z`, which is the latest version with `metis 5.1.0`.

Since `nextstrain-base` is the only package not fully-constrained by a version and build in this solving operation, it's likely the only flexibility the solver has to address dep resolution conflicts.

I'm guessing that some difference in the solver or resolution algorithm between the conda-base builds and this version of Micromamba are causing the former to produce something the latter thinks is in conflict. Since they should use broadly the same solver/algo (libmamba → libsolv), this would imply that using a newer Micromamba version might fix this.

But also, I expect pinning the `nextstrain-base` version on setup would also do the trick, and is more explicitly what we want anyway. Not doing it was kind of an oversight on my part in the https://github.com/nextstrain/cli/pull/280 work. (Very understandable oversight though!)

**Comment by tsibley** [comment]
> But also, I expect pinning the `nextstrain-base` version on setup would also do the trick, and is more explicitly what we want anyway.

…but actually is not sufficient on its own:

```
>>> micromamba("create", "-vvv", "--dry-run", "nextstrain-base ==20230830T164409Z hb0f4dca_0_locked")
…
    Encountered problems while solving:
      - package nextstrain-base-20230830T164409Z-hb0f4dca_0_locked requires suitesparse ==5.10.1 h9e50725_1, but none of the providers can be installed
    
    The environment can't be solved, aborting the operation

…
```

**Comment by tsibley** [comment]
Ok, my reading of the libsolv details in the previous comment and double checking the two suitesparse 5.10.1 packages available on conda-forge has me thinking that boa (used in conda-base builds) is producing a bad solve for the versions of suitesparse and metis. Micromamba seems correct here. (But then again, it also does the upgrade to the latest conda-base just fine?? I'm still confused by that still.)

I upgraded Micromamba to 1.5.0 (latest version), and it still doesn't like the latest package, but at least it has a better error message:

```
error    libmamba Could not solve for environment specs
    The following package could not be installed
    └─ nextstrain-base ==20230830T164409Z hb0f4dca_0_locked is not installable because it requires
       ├─ metis ==5.1.1 h59595ed_1, which can be installed;
       └─ suitesparse ==5.10.1 h9e50725_1, which requires
          └─ metis >=5.1.0,<5.1.1.0a0 , which conflicts with any installable versions previously reported.
```

This matches my reading of libsolv above.

**Comment by tsibley** [comment]
Still very confused how "install old, update to latest" works and how other CI jobs installed the latest just fine ([e.g.](https://github.com/nextstrain/conda-base/actions/runs/6027689395/job/16353600644)). This feels like something changing at a distance.

**Comment by tsibley** [comment]
I'd thought maybe Nextstrain CLI 7.2.0's relatively recent upgrade of Micromamba 1.0.0 → 1.1.0 might have been implicated, but 1.0.0 exhibits the same issues locally and besides, 7.2.0 was released 2 weeks ago, well before recent CI jobs like the one linked above passed.

**Comment by joverlee521** [comment]
> This feels like something changing at a distance.

Looks like new builds of metis 5.1.0 and 5.1.1 were [released a couple days ago](https://anaconda.org/conda-forge/metis/files), maybe some changes in dependencies there?

**Comment by tsibley** [comment]
I think I have it figured out. Don't think it's our fault. Let me confirm.

**Comment by tsibley** [comment]
Anaconda appears to have incorrect indexing of (both builds of) suitesparse 5.10.1.

Compare the metadata for the distribution (used post-install to solve deps for subsequent installs):

    $ curl https://api.anaconda.org/release/conda-forge/suitesparse/5.10.1 | jq '
    >   .distributions | map(
    >       select(.attrs.subdir == "linux-64")
    >     | [
    >       .full_name,
    >       (.attrs.depends | map(select(startswith("metis "))) | .[0])
    >     ]
    >   )
    > '
    [
      [
        "conda-forge/suitesparse/5.10.1/linux-64/suitesparse-5.10.1-h9e50725_1.tar.bz2",
        [
          "metis >=5.1.0,<5.2.0a0"
        ]
      ],
      [
        "conda-forge/suitesparse/5.10.1/linux-64/suitesparse-5.10.1-hd8046ac_0.tar.bz2",
        [
          "metis >=5.1.0,<5.2.0a0"
        ]
      ]
    ]

with the metadata in the channel index (used pre-install to solve deps):

    $ curl https://conda.anaconda.org/conda-forge/linux-64/repodata.json.zst | zstdcat | jq '
    >   .packages | to_entries | map(
    >       select(.key | contains("suitesparse-5.10.1"))
    >     | [
    >       .key,
    >       (.value.depends | map(select(startswith("metis "))))
    >     ]
    >   )
    > '
    [
      [
        "suitesparse-5.10.1-h9e50725_1.tar.bz2",
        [
          "metis >=5.1.0,<5.1.1.0a0"
        ]
      ],
      [
        "suitesparse-5.10.1-hd8046ac_0.tar.bz2",
        [
          "metis >=5.1.0,<5.1.1.0a0"
        ]
      ]
    ]

This is why initial install of `nextstrain-base ==20230830T164409Z` fails but upgrade to that same version succeeds: the former uses the channel index metadata for suitesparse, the latter the locally installed distribution metadata (e.g. `${prefix}/conda-meta/suitesparse-5.10.1-h9e50725_1.json`).

I confirmed that the distribution metadata API is indeed returning the metadata from the actual distribution:

    $ curl https://conda.anaconda.org/conda-forge/linux-64/suitesparse-5.10.1-h9e50725_1.tar.bz2 \
    > | tar -xjO info/index.json \
    > | jq '.depends | map(select(startswith("metis ")))'
    [
      "metis >=5.1.0,<5.2.0a0"
    ]

    $ curl https://conda.anaconda.org/conda-forge/linux-64/suitesparse-5.10.1-h9e50725_1.tar.bz2 \
    > | tar -xjO info/recipe/meta.yaml \
    > | yq '.requirements.run | map(select(startswith("metis ")))'
    [
      "metis >=5.1.0,<5.2.0a0",
      "metis >=5.1.0,<5.2.0a0"
    ]

and that it's all the same as what's in the local install:

    $ jq '.depends | map(select(startswith("metis ")))' "$NEXTSTRAIN_HOME"/runtimes/conda/env/conda-meta/suitesparse-5.10.1-h9e50725_1.json
    [
      "metis >=5.1.0,<5.2.0a0"
    ]

In short:

> There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors.
> —[Leon Bambrick](https://twitter.com/secretGeek/status/7269997868), via [Martin Fowler](https://www.martinfowler.com/bliki/TwoHardThings.html)

and we're hitting caching issues.  (An index is a cache.)

**Comment by tsibley** [comment]
It gets messier: the difference in the index vs. distribution metadata is not accidental, but intentional.

I went to [read about channel indexing](https://docs.conda.io/projects/conda-build/en/latest/concepts/generating-index.html#how-an-index-is-generated) and noticed this step (emphasis mine):

> For each subdir:
>
> 1. Look at all the packages that exist in the subdir.
> 2. Generate a list of packages to add/update/remove.
> 3. Remove all packages that need to be removed.
> 4. For all packages that need to be added/updated:
>     - Extract the package to access metadata, including full package name, file modification time (`mtime`), size, and `index.json`.
>      - Aggregate package metadata to repodata collection.
> 5. **Apply repodata hotfixes (patches).**
> 6. Compute and save the reduced current_index.json index.

That raised my eyebrows. So I [read further about repodata patching](https://docs.conda.io/projects/conda-build/en/latest/concepts/generating-index.html#repodata-patching), which mentioned how conda-forge applies repodata patches using https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/.

Any sign of suitesparse or metis in there? Oh, you bet!

https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/commit/2a2c288c1cfa2e73873e04478ba16576a762e829

Committed just a couple days ago. So this is intentional, to fix an actual ABI breakage, but it has the side-effect of breaking a previously-working combination of packages. This unfortunate risk is noted by the Conda docs linked to above:

> Hotfixing is tricky, as it has the potential to break environments that have worked, but it is also sometimes necessary to fix environments that are known not to work.

I think if we rebuild conda-base again, now that the hotfixing is in place, we'll be ok for new installs again. Will confirm that next.

**Comment by tsibley** [comment]
That conda-forge repodata patches change was [merged 30 Aug at about 10:19 US/Pacific](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/527#event-10235614388). To take effect it then would have to be built, uploaded, and finally used by Anaconda during index update.

Our latest nextstrain-base version (`20230830T164409Z`) starting building at 9:44 and finished by around 9:55, so wouldn't have seen the new hotfix patch to the repodata. orz

**Comment by tsibley** [comment]
Rebuild is looking promising already.

![image](https://github.com/nextstrain/monkeypox/assets/79913/aab3e955-2c3a-4e97-99dc-f12e0f062520)

**Comment by tsibley** [comment]
So assuming that rebuild mostly* resolves the issue for now, how do we avoid similar issues in the future?

One way might be having scheduled CI in conda-base that regularly tests if the latest package version is still initially installable (similar to how [Nextstrain CLI regularly tests if its standalone installers still work](https://github.com/nextstrain/cli/blob/3c93e87c2ac6bdd1b6a913070f709067b29a6cc1/.github/workflows/standalone-installers.yaml), since they're also dependent on external resources). If that test breaks, we get an early warning to see what's up. If we're really fancy, we could potentially even try to detect certain kinds of breakages like this kind here and automatic remediate it by kicking off another package build.

\* nextstrain-base versions between `(20230717T174555Z, 20230830T164409Z]` are still forever broken for initial installs, but could be upgraded to.

**Comment by tsibley** [comment]
This looks resolved by the just-released nextstrain-base `20230901T214523Z`.

**Comment by tsibley** [comment]
Closing this as this reported issue is resolved. We'd maybe like to do more to prevent it from happening in the future, but I opened a conda-base issue for that: https://github.com/nextstrain/conda-base/issues/41

**Comment by victorlin** [comment]
> https://github.com/nextstrain/monkeypox/issues/177#issuecomment-1701497622: we might still want to change `nextstrain setup conda` to use the same logic for figuring out the latest version as `nextstrain update` does rather than leaving it to Micromamba.

https://github.com/nextstrain/cli/issues/318
